### ISSUE TITLE
Randomized ETKDG Sampling

### ISSUE THREAD

**Issue by SabariKumar** [SabariKumar/confsweeper#7]
# Exhaustive randomized ETKDG for macrocyclic-peptide conformer sampling

## Background
Copy of peptide-electrostatics issue 6

confsweeper's current pipeline runs nvmolkit ETKDG at `n_confs=100` and (optionally) Pool B torsional sampling, then clusters with Butina at `cutoff_dist=0.1`. The design intent is to replace CREST metadynamics with cheap GPU randomization for cyclic peptides.

A direct CREMP comparison (36 198 macrocyclic peptides, CREST-generated ensembles) shows the current pipeline is severely under-sampling:

| metric | CREMP (CREST) | confsweeper (ETKDG, `n_confs=100`) |
|---|---|---|
| median Boltzmann weight of dominant conformer | 25.4 % | ≈ 72 % |
| peptides with `max_bw > 0.95` | 0.4 % | majority |
| mean unique conformers per peptide | 864 | 100 (only ~3 within 3 kT) |

A Butina cutoff sweep (0.1 → 0.01) on a representative PAMPA peptide kept all 100 conformers as their own clusters at every cutoff, confirming clustering is **not** the bottleneck — sampling diversity is. CREMP shows the underlying macrocyclic-peptide landscape is generally not one-hot, so the sharpness we observe is purely an artifact of ETKDG with 100 seeds finding one basin per molecule.

This directly impacts the PAMPA fine-tuning val/MAE: only one conformer per molecule contributes to the property prediction. The wrapper had to be initialised with α = 0.01 (near-uniform) just to keep gradients flowing — a workaround for one-hot ensembles, not a fix.

## Proposed approach

Keep the philosophical commitment to randomization-over-metadynamics. Scale ETKDG up by orders of magnitude, exploiting the fact that nvmolkit's GPU embed is ~100× cheaper than CREST per conformer. Pipeline (one new entry point in `confsweeper.py`):

1. **Massive embed** — nvmolkit ETKDG at N random seeds (target N picked by saturation experiments; expect 1k–10k).
2. **Batched MACE scoring** — single GPU forward pass over the full pool, chunked if memory-bound.
3. **Energy filter** — drop conformers more than `E_window` above the minimum (default ~5 kT) before clustering. High-energy conformers cannot contribute to BW regardless of geometric uniqueness, so spending Butina time on them is wasted.
4. **Energy-ranked geometric dedup** — sort survivors by MACE energy, iteratively pick the lowest-energy remaining conformer as a basin representative, exclude all geometric neighbors within RMSD threshold. Differs from vanilla Butina, which orders by cluster density rather than energy and so can drop the lowest-energy member of a dense cluster.
5. **Output** — centroid IDs + MACE energies, identical contract to existing `get_mol_PE_batched`.

Optional toggle: MMFF94 or MACE-driven local minimization of each ETKDG conformer before scoring, to push random-init structures into their basin minima. Behind a flag, default off until measured against the no-minimization baseline.

## Diagnostic experiments before committing to N

Run a saturation curve on 3–5 representative CycPeptMPDB peptides spanning small (~50 heavy atoms) / medium (~70) / large (~100+):

- N ∈ {100, 500, 1k, 5k, 10k, 50k} ETKDG seeds
- per N, report: `max_bw`, `eff_n`, `entropy`, `n_within_3kT`, wall-clock per stage
- pick N where each curve flattens; validate `max_bw` distribution lands near CREMP's median 25 %

Also re-benchmark torsional sampling (Pool B) on top of exhaustive ETKDG. If exhaustive ETKDG saturates on its own, the torsional path becomes redundant and can be deprecated.

## Key questions / unknowns

- Does `nvmolkit.embedMolecules.EmbedMolecules(..., confsPerMolecule=N)` give N independently-seeded embeddings, or does it share random state across the batch? If the latter, the pipeline calls EmbedMolecules in a loop with explicit seed offsets.
- GPU memory ceiling per single nvmolkit embed call — sets the chunk size for large N.
- Does ETKDG's internal cleanup land conformers in basin minima, or is explicit MMFF94 / MACE local minimization required to make geometric dedup meaningful?
- Does PAMPA val/MAE actually drop with rich ensembles? Ultimate test: re-fit `FinetuneJanossyWrapper` and check whether learned α moves toward 1 (it currently lives at 0.01 to compensate for one-hot inputs) and whether MAE drops.

## Relationship to other work

- **Upstream**: introduces a new `get_mol_PE_exhaustive` (or extension of `get_mol_PE_batched`) in `confsweeper.py`. Deprecates `n_confs=100` as the default for macrocyclic peptides.
- **Downstream**: full regeneration of `data/fine_tune/pampa_conformers.pkl` and `pampa_conformer_features.h5`, then re-run `runtime/train_pampa.sh`. Fine-tuning hyperparameters (LR, α init) likely need re-tuning against the new BW distribution.
- **Related**: torsional sampling becomes a candidate for deprecation if exhaustive ETKDG covers its diversity gain — re-benchmark before removing.
- **Validation**: target metric for success is CREMP-comparable `max_bw` and `n_within_3kT` distributions on PAMPA; secondary metric is PAMPA val/MAE drop after fine-tuning re-run.

## Out of scope

- Switching to CREST. The point of this issue is the opposite: hold the line on randomization-over-metadynamics by scaling ETKDG.
- Replacing MACE-OFF23 as the energy backend.
- Property-head architecture changes; this issue only concerns conformer sampling.

**Comment by SabariKumar** [comment]
<p style="white-space: pre-wrap; margin-top: 0px; margin-bottom: 0.2em; color: rgb(204, 204, 204); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(31, 31, 31); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Semantics check is conclusive:</p>

Path | Setup | conformers | unique
-- | -- | -- | --
A | single call, confsPerMolecule=200, seed=0 | 200 | 200
B | two calls, confsPerMolecule=100, seed=0 then seed=1000 | 200 | 200
C | two calls, confsPerMolecule=100, seed=0 then seed=0 | 200 | 200 (!)

<p style="white-space: pre-wrap; margin-top: 0.1em; margin-bottom: 0.2em; color: rgb(204, 204, 204); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(31, 31, 31); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Findings:</p><ol style="padding-inline-start: 2em; color: rgb(204, 204, 204); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(31, 31, 31); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li><strong>Path A is genuinely seed-independent within a single call</strong><span> </span>—<span> </span><code style="font-family: var(--app-monospace-font-family); color: var(--vscode-textPreformat-foreground); background-color: var(--vscode-textPreformat-background); padding: 2px 4px; border-radius: 3px; word-break: break-word; font-size: 0.9em;">confsPerMolecule=N</code><span> </span>gives N distinct embeddings (no internal seed sharing collapsing the batch).</li><li><strong>Path B (chunked, different seeds) covers comparable diversity</strong><span> </span>to Path A — nearest-neighbor distance distributions match (min ≈ 28–31 Å, median ≈ 43 Å on this peptide). Chunking is safe as a memory fallback.</li><li><strong>Path C is surprising</strong>: even with<span> </span><code style="font-family: var(--app-monospace-font-family); color: var(--vscode-textPreformat-foreground); background-color: var(--vscode-textPreformat-background); padding: 2px 4px; border-radius: 3px; word-break: break-word; font-size: 0.9em;">randomSeed=0</code><span> </span>repeated, the second call produced 100 different conformers from the first. So<span> </span><code style="font-family: var(--app-monospace-font-family); color: var(--vscode-textPreformat-foreground); background-color: var(--vscode-textPreformat-background); padding: 2px 4px; border-radius: 3px; word-break: break-word; font-size: 0.9em;">params.randomSeed</code><span> </span>does<span> </span><strong>not</strong><span> </span>alone make Path A bit-exact reproducible — nvmolkit advances some internal state we can't see. Implications:<ul style="padding-inline-start: 2em;"><li>Reproducibility across runs cannot rely on just setting<span> </span><code style="font-family: var(--app-monospace-font-family); color: var(--vscode-textPreformat-foreground); background-color: var(--vscode-textPreformat-background); padding: 2px 4px; border-radius: 3px; word-break: break-word; font-size: 0.9em;">randomSeed</code>. Best we can do for tests is fix<span> </span><code style="font-family: var(--app-monospace-font-family); color: var(--vscode-textPreformat-foreground); background-color: var(--vscode-textPreformat-background); padding: 2px 4px; border-radius: 3px; word-break: break-word; font-size: 0.9em;">n_seeds</code><span> </span>and assert<span> </span><strong>statistical</strong><span> </span>equivalence (sorted energies match within float tolerance), not coordinate-exact.</li><li>For the chunked-equivalence test in Phase 4, we should compare<span> </span><em>sorted-energy distributions</em><span> </span>and<span> </span><em>centroid count distributions</em>, not coordinate identity.</li></ul></li></ol><br class="Apple-interchange-newline">
